src/SignalR/clients/csharp/Client.SourceGenerator/src/CallbackRegistrationGenerator.Parser.cs

```diff
@@ -0,0 +1,128 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNetCore.SignalR.Client.SourceGenerator
+{
+    internal partial class CallbackRegistrationGenerator
+    {
+        public class Parser
+        {
+            private readonly GeneratorExecutionContext _context;
+
+            public Parser(GeneratorExecutionContext context)
+            {
+                _context = context;
+            }
+
+            public SourceGenerationSpec Parse(List<MemberAccessExpressionSyntax> syntaxList)
+            {
+                var sourceGenerationSpec = new SourceGenerationSpec();
+                var compilation = _context.Compilation;
+
+                var providerSymbols = new Dictionary<string, (ITypeSymbol, MemberAccessExpressionSyntax)>();
+                var iHubConnectionType =
+                    compilation.GetTypeByMetadataName("Microsoft.AspNetCore.SignalR.Client.IHubConnection");
+
+                // Go thru candidates and filter
+                foreach(var memberAccess in syntaxList)
+                {
+                    var expressionModel = _context.Compilation.GetSemanticModel(memberAccess.Expression.SyntaxTree);
+                    var typeInfo = expressionModel.GetTypeInfo(memberAccess.Expression);
+
+                    // Filter based on receiver symbol
+                    if(!SymbolEqualityComparer.Default.Equals(typeInfo.Type, iHubConnectionType) &&
+                       (typeInfo.Type.AllInterfaces.IsDefaultOrEmpty ||
+                        !typeInfo.Type.AllInterfaces.Any(x => SymbolEqualityComparer.Default.Equals(x, iHubConnectionType))))
+                    {
+                        // Member access is not acting on IHubConnection or a type implementing it; as such we will skip
+                        continue;
+                    }
+
+                    // Extract type symbol
+                    ITypeSymbol symbol;
+                    if (memberAccess.Name is GenericNameSyntax {Arity: 1} gns)
```

Comment:
I was confused as to how this was finding `RegisterCallbackProvider` at first, then saw that the candidates were filtered and passed in. Maybe add a comment at the top of `Parse` saying that the list is already filtered 😄

Reply:
I added more comments but I still feel some portions remain a bit intimidating especially in emitter. Definitely needs one more pass. It'd be helpful if more eyes can pinpoint specific areas as lacking :)